File: interviewProject/src/interview/sort/QuickSort.java
package interview.sort;

import static java.lang.System.out;

/**
 * Created by zhao on 12/11/16.
 */
public class QuickSort {

    private int[] input;

    public static void main(String[] args) {
        int[] in = new int[]{6, 5, 3, 1, 8, 7, 2, 4};
        QuickSort qs = new QuickSort();
        qs.sort(in);
        for (int i : in) {
            out.print(i + " ");
        }

    }

    public void sort(int[] numbers) {
        if (numbers == null || numbers.length == 0) {
            return;
        }
        this.input = numbers;
        quickSort(0, numbers.length - 1);
    }

    public void quickSort(int low, int high) {
        int i = low;
        int j = high;
        int pivot = input[low + (high - low) / 2];
        while (i <= j) {
            while (input[i] < pivot) {
                i++;
            }
            while (input[j] > pivot) {
                j--;
            }
            if (i <= j) {
                swap(i, j);
                i++;
                j--;
            }
        }
        if (low < j) {
            quickSort(low, j);
        }
        if (i < high) {
            quickSort(i, high);
        }

    }

    private void swap(int i, int j) {
        int temp = input[i];
        input[i] = input[j];
        input[j] = temp;
    }

}

File: funtionalprogramming/src/main/scala/algorithm/stfd/QuickSort.scala
package algorithm.stfd

/**
  * Created by zhao on 2/7/18.
  */
object QuickSort {

  def main(args: Array[String]): Unit = {
    val input = Array(1, 3, 5, 7, 9, 2, 4, 6, 8, 10)
    val sortedInput = sort(input, 0, input.length - 1)
    sortedInput.foreach(x => print(s"$x,"))
  }

  def partition(input: Array[Int], l: Int, r: Int): (Array[Int], Int) = {
    var i = l + 1
    val pivot = input(l)
    for (j <- l + 1 to r) {
      if (input(j) < pivot) {
        val t = input(j)
        input(j) = input(i)
        input(i) = t
        i += 1
      }
    }
    val t = input(i - 1)
    input(i - 1) = pivot
    input(l) = t
    (input, i - 1)
  }

  def sort(input: Array[Int], l: Int, r: Int): Array[Int] = {
    if (l >= r) return input
    val pivotIdx = choosePivot(input, l, r)
    //Put pivot as first element in array
    val t = input(l)
    input(l) = input(pivotIdx)
    input(pivotIdx) = t
    val part = partition(input, l, r)
    sort(part._1, l, part._2 - 1)
    sort(part._1, part._2 + 1, r)


  }

  def choosePivot(input: Array[Int], i: Int, r: Int): Int = i

}